org.eclipse.scout.sdk.s2i/src/main/kotlin/org/eclipse/scout/sdk/s2i/util/SourceFolderHelper.kt

```diff
@@ -106,122 +98,210 @@ class SourceFolderHelper(val project: Project, val sourceFolder: SourceFolder, v
-         */
-        fun findServerSourceFolderPair(project: Project, sharedSourceFolder: SourceFolder?, scoutTierOfModule: ((Module) -> ScoutTier?)? = null, findClasspathEntry: (VirtualFile?) -> IClasspathEntry?): Pair<SourceFolder, IClasspathEntry>? {
-            if (sharedSourceFolder == null) {
-                return null
+        m_tierSourceFolderBundleMap[tier]?.let { return it }
+
+        // consider a tree like this, uppercase letters represent nodes for whom we already computed the sourceFolderBundle
+        //
+        //    a <- root
+        //   b
+        //  c d
+        // E   f <- example 1
+        //  G
+        //   h
+        //    i <- example 2
+
+        // example 1
+        // search for node f, which is not a descendant of any node for whom we already computed the sourceFolderBundle
+        // -> we need to compute some ancestors first before we can compute the descendants down to f (after all necessary descendants are computed this example is equivalent to example 2)
+
+        // example 2
+        // search for node i, which is a descendant of at least one node for whom we already computed the sourceFolderBundle
+        // -> we just need to compute some more descendants
+
+        // get a path (start inclusive, end exclusive) from the root node to the first node for whom we already computed the sourceFolderBundle
+
+        // example 1 & 2: E
+        var existingTier: ITier<*>? = null
+        // example 1 & 2: a > b > c
+        val pathToExistingTier = TierTree.path(null, m_tierSourceFolderBundleMap.first().key)
+            .takeWhile { t ->
+                existingTier = t
+                return@takeWhile !m_tierSourceFolderBundleMap.containsKey(t)
+            }
+
+        // get a path (start exclusive, end inclusive) from the last node for whom we already computed the sourceFolderBundle or that is contained in the path we just created to the node we search for
+
+        // example 1: b
+        // example 2: G
+        var startTier: ITier<*>? = null
+        // example 1: d > f
+        // example 2: h > i
+        val pathToTier = TierTree.path(null, tier).reversed()
+            .takeWhile { t ->
+                startTier = t
+                return@takeWhile !m_tierSourceFolderBundleMap.containsKey(t) && !pathToExistingTier.contains(t)
             }
-            return findClosestDependentSourceFolder(project, ScoutTier.Server, false, sharedSourceFolder, scoutTierOfModule, findClasspathEntry)
+            .reversed()
+
+        if (pathToExistingTier.contains(startTier)) {
+            // we need to compute sourceFolderBundles for all elements from the end of pathToExistingTier until startTier
+            pathToExistingTier
+                .subList(pathToExistingTier.lastIndexOf(startTier), pathToExistingTier.size) // example 1: b > c
```

Comment:
Instead of creating a sublist and then reverse, it would also be possible to just make a loop from pathToExistingTier.size-1 to pathToExistingTier.lastIndexOf(startTier)? Then those lists are not necessary?